src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java

```diff
@@ -252,29 +233,37 @@ public void setMaximumPoolingConnections(int maximumPoolingConnections) {
     }
   }
 
-  /**
-   * (Re)initializes the HttpContext object. This removes any existing state (mainly cookies). Use an own
-   * cookie store, instead of the httpClient's global store, so cookies get reset on reinitialization
-   */
-  private void initializeHttpContext() {
-    CookieStore cookieStore = new BasicCookieStore();
-    httpContext = HttpClientContext.create();
+  protected HttpClientContext createHttpClientContext() {
+	HttpClientContext httpContext = HttpClientContext.create();
     httpContext.setCookieStore(cookieStore);
+    return httpContext;
   }
 
   @Override
   public void close() {
-    try {
-      httpClient.close();
-    } catch (IOException e) {
-      LOG.debug(e);
+    if(httpClient!=null) {
+	  synchronized(httpClientLock){
+	    if(httpClient!=null) {
```

Comment:
Sure, I can do that. But wouldn't it be better to do that in another pull request?

I'm happy to do it here, but to me, it seems like this pull request should solve one issue (thread safety) and not expand to other issues (like this kind of refactoring/cleanup). Let me know.